kvserver: track priority inversion in replicate queue metrics

wenyihu6 opened the pull request:
Part of: https://github.com/cockroachdb/cockroach/issues/151847
Resolves: https://github.com/cockroachdb/cockroach/issues/152022
Release note: none 

----


**kvserver: track priority inversion in replicate queue metrics**

Previously, replicas could be enqueued at a high priority but end up processing
a lower-priority actions, causing priority inversion and unfairness to other
replicas behind them that needs a repair action. This commit adds metrics to
track such cases. In addition, this commit also adds metrics to track when
replicas are requeued in the replicate queue due to a priority inversion from a
repair action to a rebalance action.

---

**kvserver: add TestPriorityInversionRequeue**

Previously, we added priority inversion requeuing mechanism.

This commit adds a unit test that forces the race condition we suspected to be
happening in escalations involving priority inversion and asserts that priority
inversion occurs and that the replica is correctly requeued. Test set up:

1. range’s leaseholder replica is rebalanced from one store to another.
2. new leaseholder enqueues the replica for repair with high priority (e.g. to
finalize the atomic replication change or remove a learner replica)
3. before processing, the old leaseholder completes the change (exits the joint
   config or removes the learner).
4. when the new leaseholder processes the replica, it computes a
   ConsiderRebalance action, resulting in a priority inversion and potentially
   blocking other high-priority work.

---

**kvserver: delete per action priority inversion metrics**

This commit removes per-action priority inversion metrics due to their high
cardinality. We already have logging in place, which should provide sufficient
observability. For now, we care about is priority inversion that leads to
consider rebalance and requeuing the most.

wenyihu6 committed:
kvserver: track priority inversion in replicate queue metrics

Previously, replicas could be enqueued at a high priority but end up processing
a lower-priority actions, causing priority inversion and unfairness to other
replicas behind them that needs a repair action. This commit adds metrics to
track such cases. In addition, this commit also adds metrics to track when
replicas are requeued in the replicate queue due to a priority inversion from a
repair action to a rebalance action.

wenyihu6 committed:
kvserver: add TestPriorityInversionRequeue

Previously, we added priority inversion requeuing mechanism.

This commit adds a unit test that forces the race condition we suspected to be
happening in escalations involving priority inversion and asserts that priority
inversion occurs and that the replica is correctly requeued. Test set up:

1. range’s leaseholder replica is rebalanced from one store to another.
2. new leaseholder enqueues the replica for repair with high priority (e.g. to
finalize the atomic replication change or remove a learner replica)
3. before processing, the old leaseholder completes the change (exits the joint
   config or removes the learner).
4. when the new leaseholder processes the replica, it computes a
   ConsiderRebalance action, resulting in a priority inversion and potentially
   blocking other high-priority work.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152697)
<!-- Reviewable:end -->

wenyihu6 commented:
@sumeerbhola re: the original comments you had on https://github.com/cockroachdb/cockroach/pull/152624#pullrequestreview-3164690636.

TFTR! 

> labeled metrics 

Agree that the high cardinality is unfortunate. But I'm not sure whether there is a better solution. We need to backport this to v24.3. It doesn’t look like it was included yet (https://github.com/cockroachdb/cockroach/pull/143536). I’m also not sure if `tsdump` would suport this feature. We can consider improving this only on master in a follow up pr. For now, I'm leaning towards taking the loss on high cardinality unless we are also interested in something around tracking the priority distribution in the priority queue. Wdyt?

> enqueue counter metric

We only have metrics for measuring pending queue length for now https://github.com/cockroachdb/cockroach/blob/364231907b129337ae16fc2e1f93c5defaaba395/pkg/kv/kvserver/queue.go#L138. It's a gauge. I plan to add something for enqueue success and enqueue failures in a follow up pr.

wenyihu6 commented:
I’m adding a new test, TestPriorityInversionRequeue, to try to force the suspected race condition. It fails under stress on CI. I checked the logs, and it was because if n1 removes the learner too quickly, n4 may miss it when enqueuing to the replicate queue. I’m not too concerned, since the goal is mainly to reproduce and confirm that the race could occur. Things pass under stress for me locally. We might need another testing knob. Do you think this test is worth adding? If so, I’ll continue exploring ways to delay n1’s learner removal until n4 has a chance to enqueue.

tbg reviewed (changes requested):
Basically looks good, but let's sync on whether we really meant to add the metrics tracking inversions on a per-type basis.
Thank you for adding the test - nice to see the mechanism confirmed :fire: 

> @tbg reviewed 1 of 1 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 2 of 3 files at r4, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-:-OZ3VEYY0PikLfdi_Wgm:b24x1c3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/replicate_queue_test.go` line 2564 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-OZ3UFhh1Zw2T8kAJQvF:-OZ3UFhh1Zw2T8kAJQvG:blw95ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/2471a486850805305cc6814454fc01857c26d2c7/pkg/kv/kvserver/replicate_queue_test.go#L2564)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 	defer log.Scope(t).Close(t)
> ```

I'd suggest you skip this test under duress preemptively.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 322 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-OZ3UkpCADg_gc1_lrqn:-OZ3UkpCADg_gc1_lrqo:b-dscuse) ([raw file](https://github.com/cockroachdb/cockroach/blob/2471a486850805305cc6814454fc01857c26d2c7/pkg/kv/kvserver/replicate_queue.go#L322)):*
> ```Go
> 		Unit:        metric.Unit_COUNT,
> 	}
> 	metaReplicateQueuePriorityInversionForAddVoterCount = metric.Metadata{
> ```

I thought we discussed that given that we have (rate-limited) logging and total counters, we could do away with the "per-inversion-type" counters. Am I misremembering this?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: delete per action priority inversion metrics

This commit removes per-action priority inversion metrics due to their high
cardinality. We already have logging in place, which should provide sufficient
observability. For now, we care about is priority inversion that leads to
consider rebalance and requeuing the most.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 322 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-OZ3UkpCADg_gc1_lrqn:-OZ6kuC09y1foOcl8Dll:b-rnlpfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2471a486850805305cc6814454fc01857c26d2c7/pkg/kv/kvserver/replicate_queue.go#L322)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I thought we discussed that given that we have (rate-limited) logging and total counters, we could do away with the "per-inversion-type" counters. Am I misremembering this?
</blockquote></details>

:disappointed: Don’t recall reaching a conclusion in that discussion. Might have missed some discussion points. I’ve removed it for now.
___
*[`pkg/kv/kvserver/replicate_queue_test.go` line 2564 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-OZ3UFhh1Zw2T8kAJQvF:-OZ6lSl0EqgpuNurI6LZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2471a486850805305cc6814454fc01857c26d2c7/pkg/kv/kvserver/replicate_queue_test.go#L2564)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'd suggest you skip this test under duress preemptively.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Not approving yet, but approved once unstacked.  :thumbsup: 

> @tbg reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-:-OZ878dd8sEoJE_M42Xc:b-swzzip)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
The last two pushes fixed a rebase mess-up and a linter warning.

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @tbg reviewed 7 of 7 files at r9, 2 of 5 files at r10, 3 of 3 files at r12, 3 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152697#-:-OZIhL5w2YqB0oLc5xDU:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Rebased on master to resolve merge conflicts caused by related prs.
